### ISSUE TITLE
Proposal: allow field formal parameters in setters

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/language#217]
@dmincu had this idea in an AngularDart training and I think it makes perfect sense. Not sure if this has been considered before, it seems like an easy non-breaking change to add.

This would be useful for cases where you override a setter that has a private backing store where the side-effect is not a special assignment of that backing property:

```dart
class C {
  int _myInt;

  set myInt(this._myInt) {
    print("set myInt to $_myInt");
  }

  // vs
  set myInt(int myInt) {
    _myInt = myInt;
    print("set myInt to $_myInt");
  }
  ...
}
```

as well as cases where you override the setter only because you overrode the getter (and still want both for your field):

```dart
class C {
  int _myInt;

  set myInt(this._myInt); // nothing special
  int get myInt {
    // ...
  }
}
```

I wouldn't go so far as to file these as language issues, however, I think the feature is still something that would be useful.

**Comment by tatumizer** [comment]
Setter has a formal parameter right now. 
```
class Foo {
  int _value;
  set value(int v) {
    _value = v; // you can do whatever you want here, unrelated to _value
  }
}
void main() {
   new Foo()..value=5;
}
```
Probably, you meant something else?

**Comment by munificent** [comment]
I think this is a great idea. In fact, I'd take it farther and allow *any* parameter in any method to be an initializing (well, assigning) formal.

**Comment by andreashaese** [comment]
Looks like a very fitting addition to Dart 👍 

On a tangential note, I would expect the following to work too:
```dart
get myInt => _myInt;
```
Interestingly, `myInt`'s type can't be inferred (`dynamic` in non-strong mode, error in strong mode). Does that justify a separate pitch?

**Comment by tatumizer** [comment]
dart 2 has only strong mode, nothing else. The form `get myInt => _myInt;` works fine if _myInt is declared as int, you can verify it on https://dartpad.dartlang.org

**Comment by andreashaese** [comment]
> dart 2 has only strong mode, nothing else

Sorry, I should have been more clear: I was referring to the analyzer's [explicit strong mode](https://www.dartlang.org/guides/language/analysis-options#enabling-additional-type-checks) that emits errors instead of defaulting to `dynamic` if a type can't be inferred during static analysis:

```yaml
analyzer:
  strong-mode:
    implicit-casts: false
    implicit-dynamic: false
```

> if _myInt is declared as int, the type gets inferred correctly

Are you using a secret Dart 3 version of DartPad 😄 Or do you mean the runtime type? If I copy&paste the following code into DartPad
```dart
class Test {
  int _myInt;
  get thisIsInt => _myInt;
  get thisIsBool => true;
}
```
and place the cursor on `thisIsInt` and `thisIsBool`, their static types are shown as `get thisIsInt → dynamic` and `get thisIsBool → dynamic`.

**Comment by tatumizer** [comment]
Oh, I see. In that case, the effect is not limited to getters. You mean, for getters, there should be a special case of inference? Not sure about it. What is "getter" after all? Just a function with a different syntax.
It's not different from
```
class Test {
  foo() => 5; // type: () -> dynamic
}
```

**Comment by andreashaese** [comment]
Yes, I know. The motivation for my question is that if we get this shorthand syntax for write-only setters, then read-only getters look a bit unbalanced:
```dart
set myInt(this._myInt)
int get myInt => _myInt; // why do I need to repeat the type here?
```

Of course I wouldn't want to make an exception only for getters, but to me that would seem sensible for any method/function where the available type information is sufficient. I wonder if the current design is deliberate, given that Dart is technically able to infer unambiguous function signatures:
```dart
class Test {
  final one = (int a) => '{a}'; // (int) -> String
  two(int a) => '{a}'; // (int) -> dynamic, why?
}
```

With that said, I don't want to derail the conversation about setters and would be happy to move this over to separate issue.

**Comment by lrhn** [comment]
Late comment:

This raises a question.

An "initializing formal" in an initializing constructor is _initializing_, assigning a value directly to the storage location of the variable and bypassing any setter. You can only do that in the class declaration which introduces that variable as having a storage location, the declaring that _needs_ to initialize the variable in all initializing constructors.

If `this._foo` is used as an instance member parameter, should it directly set the storage location and bypass any overriding setter, or should it just call the setter of the target variable as a normal virtual call? 

Setting the storage directly would definitely only be available to the class declaration that introduces the variable, all subclasses can rely on is that there is a getter and setter, not that it's an instance variable. 

The constructor initializing the variable is the first time anyone touches the variable's storage, and it happens before anyone has a reference to the object. After that, an overriding getter/setter may assume all access goes through it, and someone replacing the value without going through the setter may cause logical inconsistencies.

On the other hand, it's already possible for a superclass to change a variable without going through a subclass setter by using `super.x = v`. If you want to bypass an overriding setter, you can. (But you probably shouldn't, so maybe we shouldn't make it easy.)

If `this.x` just calls the setter normally, it's not really the same operation. Maybe the syntax shouldn't be the same.

So not convinced this is an obviously great idea, not because I haven't wanted it myself, but because the sharp edge cases.
